Bitbucket.Authentication/OAuth/SimpleServer.cs

```diff
@@ -0,0 +1,124 @@
+﻿/**** Git Credential Manager for Windows ****
+ *
+ * Copyright (c) Atlassian
+ * All rights reserved.
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the """"Software""""), to deal
+ * in the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
+ * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE."
+**/
+
+using Atlassian.Shared.Authentication.Helpers;
+using System;
+using System.IO;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows;
+
+namespace Atlassian.Bitbucket.Authentication.OAuth
+{
+    /// <summary>
+    ///     Implements a simple HTTP server capable of handling Bitbucket OAuth callback requests.
+    /// </summary>
+    public class SimpleServer
+    {
+        /// <summary>
+        /// Async wait for an URL with a timeout
+        /// </summary>
+        /// <param name="url"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="RemoteHostException">Throws when there's a timeout, or other error</exception>
+        public static async Task<string> WaitForURLAsync(string url, CancellationToken cancellationToken)
+        {
+            var listener = new HttpListener { Prefixes = { url } };
+            listener.Start();
+
+            string rawUrl = "";
+            try
+            {
+                var context = await listener.GetContextAsync().RunWithCancellation(cancellationToken);
+                rawUrl = context.Request.RawUrl;
+
+                await Task.Delay(100); //Wait 100ms before writing the response out
```

Comment:
Just for reference, there's no promise made my the CLR that the `Task.Delay(100)` ever gets scheduled, additionally there will be a delay before it is, so the `100` is "just a suggestion" at best.

Why the arbitrary sleep here?